src/brs/grpc/handlers/SubmitNonceHandler.java

```diff
@@ -0,0 +1,96 @@
+package brs.grpc.handlers;
+
+import brs.Account;
+import brs.Blockchain;
+import brs.Generator;
+import brs.crypto.Crypto;
+import brs.grpc.GrpcApiHandler;
+import brs.grpc.proto.ApiException;
+import brs.grpc.proto.BrsApi;
+import brs.services.AccountService;
+
+import java.util.Objects;
+
+public class SubmitNonceHandler implements GrpcApiHandler<BrsApi.SubmitNonceRequest, BrsApi.SubmitNonceResponse> {
+
+    private final Blockchain blockchain;
+    private final AccountService accountService;
+    private final Generator generator;
+
+    public SubmitNonceHandler(Blockchain blockchain, AccountService accountService, Generator generator) {
+        this.blockchain = blockchain;
+        this.accountService = accountService;
+        this.generator = generator;
+    }
+
+    @Override
+    public BrsApi.SubmitNonceResponse handleRequest(BrsApi.SubmitNonceRequest request) throws Exception {
+        String secret = request.getSecretPhrase();
+        long nonce = request.getNonce();
+        long accountId = request.getAccount();
+        int submissionHeight = request.getBlockHeight();
+
+        if (submissionHeight != 0) {
+            if (submissionHeight != blockchain.getHeight() + 1) {
+                throw new ApiException("Given block height does not match current blockchain height");
+            }
+        }
+
+        if (Objects.equals(secret, "")) {
+            throw new ApiException("Missing Passphrase");
+        }
+
+        byte[] secretPublicKey = Crypto.getPublicKey(secret);
+        Account secretAccount = accountService.getAccount(secretPublicKey);
+        if(secretAccount != null) {
+            Account genAccount;
```

Comment:
I tend to extract larger blocks in separate functions with crisp names to improve readability